apps/electron/layers/main/src/handlers/db/sqlite.ts

```diff
@@ -33,75 +34,100 @@ interface BlobRow {
   timestamp: string;
 }
 
-export class WorkspaceDatabase {
-  sqliteDB: Database;
+const SQLITE_ORIGIN = Symbol('sqlite-origin');
+
+export class WorkspaceSQLiteDB {
+  db: Database;
   ydoc = new Y.Doc();
   firstConnect = false;
+  lastUpdateTime = ts();
 
-  constructor(public path: string) {
-    this.sqliteDB = this.reconnectDB();
+  constructor(public path: string, public workspaceId: string) {
+    this.db = this.reconnectDB();
   }
 
   // release resources
   destroy = () => {
-    this.sqliteDB?.close();
+    this.db?.close();
     this.ydoc.destroy();
   };
 
+  getWorkspaceName = () => {
+    return this.ydoc.getMap('space:meta').get('name') as string;
+  };
+
   reconnectDB = () => {
-    logger.log('open db', this.path);
-    if (this.sqliteDB) {
-      this.sqliteDB.close();
+    logger.log('open db', this.workspaceId);
+    if (this.db) {
+      this.db.close();
     }
 
     // use cached version?
-    const db = (this.sqliteDB = sqlite(this.path));
+    const db = (this.db = sqlite(this.path));
     db.exec(schemas.join(';'));
 
     if (!this.firstConnect) {
-      this.ydoc.on('update', this.addUpdateToSQLite);
+      this.ydoc.on('update', (update: Uint8Array, origin) => {
```

Comment:
what about `off`?

Reply:
I think the listener will be gc'ed after `ydoc` is destoryed, thus there is no need to unlisten it expclitly.